gear-programs/vft-treasury/src/services/mod.rs

```diff
@@ -0,0 +1,422 @@
+            gas_for_reply_deposit,
+            gas_to_send_request_to_builtin,
+            reply_timeout,
+            gas_for_transfer_to_eth_msg,
+        }
+    }
+}
+
+#[derive(Debug, Decode, Encode, TypeInfo)]
+pub struct InitConfig {
+    pub receiver_contract_address: H160,
+    pub gear_bridge_builtin: ActorId,
+    pub ethereum_event_client: ActorId,
+    pub bridging_payment_service: ActorId,
+    pub config: Config,
+}
+
+impl InitConfig {
+    pub fn new(
+        receiver_contract_address: H160,
+        gear_bridge_builtin: ActorId,
+        ethereum_event_client: ActorId,
+        bridging_payment_service: ActorId,
+        config: Config,
+    ) -> Self {
+        Self {
+            receiver_contract_address,
+            gear_bridge_builtin,
+            ethereum_event_client,
+            bridging_payment_service,
+            config,
+        }
+    }
+}
+impl<T> VftTreasury<T>
+where
+    T: ExecContext,
+{
+    pub fn seed(config: InitConfig, exec_context: T) {
+        unsafe {
+            DATA = Some(VftTreasuryData {
+                receiver_contract_address: config.receiver_contract_address,
+                gear_bridge_builtin: config.gear_bridge_builtin,
+                ethereum_event_client: config.ethereum_event_client,
+                admin: exec_context.actor_id(),
+                eth_to_vara_token_id: HashMap::new(),
+                vara_to_eth_token_id: HashMap::new(),
+                bridging_payment_service: config.bridging_payment_service,
+            });
+            CONFIG = Some(config.config);
+            MSG_TRACKER = Some(MessageTracker::default());
+        }
+    }
+
+    pub fn new(exec_context: T) -> Self {
+        Self { exec_context }
+    }
+
+    fn data(&self) -> &VftTreasuryData {
+        unsafe { DATA.as_ref().expect("VftTreasury::seed() must be called") }
+    }
+
+    fn data_mut(&mut self) -> &mut VftTreasuryData {
+        unsafe { DATA.as_mut().expect("VftTreasury::seed() must be called") }
+    }
+
+    fn config(&self) -> &Config {
+        unsafe {
+            CONFIG
+                .as_ref()
+                .expect("VftTreasury::seed() must be invoked")
+        }
+    }
+
+    fn get_eth_token_id(&self, vara_token_id: &ActorId) -> Result<H160, Error> {
+        self.data()
+            .vara_to_eth_token_id
+            .get(vara_token_id)
+            .cloned()
+            .ok_or(Error::NoCorrespondingEthAddress)
+    }
+}
+
+#[derive(Encode, Decode, TypeInfo)]
+pub enum VftTreasuryEvents {
+    Deposit {
+        from: ActorId,
+        to: H160,
+        token: ActorId,
+        amount: U256,
+    },
+    Withdraw {
+        receiver: ActorId,
+
+        token: ActorId,
+        amount: U256,
+    },
+}
+
+#[service(events = VftTreasuryEvents)]
+impl<T> VftTreasury<T>
+where
+    T: ExecContext,
+{
+    pub fn update_config(&mut self, config: Config) {
+        if self.data().admin != self.exec_context.actor_id() {
+            panic!("not admin");
+        }
```

Comment:
Maybe some method `ensure_admin`?

Reply:
@mertwole how would it be different? Should `admin` be `Option<ActorId>` in that case?

Reply:
I think the logic will be a little more clear